### ISSUE TITLE
S2583 - False positive when using null propagation followed by null coalesence

### ISSUE THREAD

**Issue by rjgotten** [SonarSource/sonar-dotnet#2134]
### Description

`S2583 - Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.` is raised incorrectly when guard conditions contain null propagation followed by null coalescence.

### Repro steps

Run analysis on code which uses null propagation followed by null coalescence in a guard condition, similar to the following:

```csharp
if (somePotentiallyNullinstance?.BooleanProperty ?? false)
{
}
```

### Expected behavior

S2583 is not raised.

### Actual behavior

S2538 is raised.

### Known workarounds

Change source code to use `!= true`.

### Related information

SonarC# Version: 7.9 (build 7583)

**Comment by valhristov** [comment]
Thanks for the feedback @rjgotten. This a known issue, described in more detail here #1347.

**Comment by rjgotten** [comment]
> especially for longer expressions

Really? Because I'd claim the inverse of that. 

Longer property chains evaluated through AND-composition, such as:

```csharp
if (foo != null && foo.Bar != null && foo.Bar.Baz != null && foo.Bar.Baz.IsTrue)
```

give quite poor readability compared to shorthand conditional access such as:

```csharp
if (foo?.Bar?.Baz?.IsTrue ?? false)
```

Personally I prefer an explicit `!= true` over the `?? false` fallback though:

```csharp
if (foo?.Bar?.Baz?.IsTrue != true)
```

As for the reading order argument: "when foo is `null` skip the remainder of this expression" doesn't seem that big of a cognitive load. It seems harder to keep track of multiple conditional expressions being AND-composed, especially when there's a risk there for continuation of the evaluation of the property chain to end up being mixed with other conditional checks.


> it is not easy to fix and we cannot promise a quick solution

That's a bit disappointing, but atleast we can work around this by creating awareness about this shortcoming in our development team, and simply marking these Sonar findings as false positives.

**Comment by andrei-epure-sonarsource** [comment]
fixed in #2497
